query-engine/query-engine-c-abi/src/engine.rs

```diff
@@ -0,0 +1,665 @@
+        let engine_id = get_cstr_safe(constructor_options.id).expect("engine id cannot be missing");
+        let log_callback_c = constructor_options.log_callback;
+        let log_callback = move |msg: String| {
+            let id = CString::new(engine_id.clone()).unwrap();
+            let c_message = CString::new(msg).unwrap();
+            unsafe {
+                log_callback_c(id.as_ptr(), c_message.as_ptr());
+            }
+        };
+
+        let str_env = get_cstr_safe(constructor_options.env).expect("Environment missing");
+        let json_env = serde_json::from_str(str_env.as_str()).expect("Environment cannot be parsed");
+        let env = stringify_env_values(json_env)?; // we cannot trust anything JS sends us from process.env
+
+        let str_datasource_overrides =
+            get_cstr_safe(constructor_options.datasource_overrides).expect("Datesource overrides missing");
+        let json_datasource_overrides =
+            serde_json::from_str(str_datasource_overrides.as_str()).expect("Datasource overrides cannot be parsed");
+        let overrides: Vec<(_, _)> = stringify_env_values(json_datasource_overrides)
+            .unwrap()
+            .into_iter()
+            .collect();
+
+        let datamodel = get_cstr_safe(constructor_options.datamodel).expect("Datamodel must be present");
+        let mut schema = psl::validate(datamodel.into());
+        let config = &mut schema.configuration;
+
+        schema
+            .diagnostics
+            .to_result()
+            .map_err(|err| ApiError::conversion(err, schema.db.source()))?;
+
+        let base_path = get_cstr_safe(constructor_options.base_path);
+        match &base_path {
+            Some(path) => env::set_current_dir(Path::new(&path)).expect("Could not change directory"),
+            _ => println!("No base path passed!"),
+        }
+
+        config
+            .resolve_datasource_urls_query_engine(
+                &overrides,
+                |key| env.get(key).map(ToString::to_string),
+                // constructor_options.ignore_env_var_errors,
+                true,
+            )
+            .map_err(|err| ApiError::conversion(err, schema.db.source()))?;
+
+        config
+            .validate_that_one_datasource_is_provided()
+            .map_err(|errors| ApiError::conversion(errors, schema.db.source()))?;
+
+        let engine_protocol = EngineProtocol::Json;
+
+        let config_dir_string = get_cstr_safe(constructor_options.native.config_dir).expect("Config dir is expected");
+        let config_dir = PathBuf::from(config_dir_string);
+
+        let builder = EngineBuilder {
+            schema: Arc::new(schema),
+            engine_protocol,
+            native: EngineBuilderNative { config_dir, env },
+        };
+
+        let log_level_string = get_cstr_safe(constructor_options.log_level).unwrap();
+        let log_level = log_level_string.parse::<LevelFilter>().unwrap();
```

Comment:
Continuing the [conversation](https://github.com/prisma/prisma-engines/pull/4774#discussion_r1530823939) about using the underlying enum, instead of parsing a string.

I think there is a tiny misunderstanding here. cbindgen will not generate headers for everything, only for the crate and since that enum comes from a dependency it is not generating it.

Reply:
Makes sense. Let's keep it the way it is 👍